### PR TITLE
Passphrase dialog support for keyrings protected by a passphrase

### PR DESCRIPTION
**Pull request by paninaro**
GUI support for when the daemon is using a passphrase-protected keyring (Linux at the moment). Includes migration UI for migrating keys from cryptfile_pass.cfg to keyring.yaml.

On launch, if the daemon's keyring is locked, the GUI will present a passphrase dialog to unlock the keyring. Additionally, a new Settings page has been added to support changing/removing/adding a keyring passphrase.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+LOCAL_TEST=true
```
Comment (paninaro) [inline .env:1]: Will remove

File: src/middleware/middleware.js [hunk 2]
```diff
@@ -62,6 +59,7 @@ const socketMiddleware = () => {
 
   const onMessage = (store) => (event) => {
     const payload = JSON.parse(event.data);
+    console.log("***** received payload: ", payload);
```
Comment (paninaro) [inline src/middleware/middleware.js:62]: Will remove

File: src/middleware/middleware_api.jsx [hunk 3]
```diff
@@ -226,6 +226,14 @@ export const handle_message = async (store, payload, errorProcessed) => {
         store.dispatch(getPlotDirectories());
       }
     }
+  } else if ((payload.command === 'keyring_status') || (payload.command === 'keyring_status_changed')) {
+    if (payload.data.success) {
+      const { is_keyring_locked } = payload.data;
+      if (is_keyring_locked == false) {
+        console.log("Keyring is unlocked, refreshing all state");
```
Comment (paninaro) [inline src/middleware/middleware_api.jsx:233]: Will remove

File: src/modules/incoming.ts [hunk 4]
```diff
@@ -235,7 +235,7 @@ export default function incomingReducer(
           logged_in: false,
           selected_fingerprint: undefined,
           public_key_fingerprints: [],
-          logged_in_received: true,
+          logged_in_received: true,
```
Comment (paninaro) [inline src/modules/incoming.ts:238]: Will remove

File: src/modules/keyring.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        ...state,
+        migration_skipped: action.skip,
+      };
+    case 'INCOMING_MESSAGE':
+      const { message } = action;
+      const { data } = message;
+      const { command } = message;
+      if ((command === 'keyring_status') || (command === 'keyring_status_changed')) {
+        if (data.success) {
+          console.log(command);
```
Comment (paninaro) [inline src/modules/keyring.ts:43]: Will remove

File: src/middleware/middleware_api.jsx (hunk 3 above)
Comment (paninaro) [reply src/middleware/middleware_api.jsx:233]: Decided to leave it intact.

File: src/components/app/App.tsx [hunk 6]
```diff
@@ -15,6 +15,7 @@ import { exit_and_close } from '../../modules/message';
 import useLocale from '../../hooks/useLocale';
 import AppModalDialogs from './AppModalDialogs';
 import AppLoading from './AppLoading';
+import AppPassLogin from './AppPassLogin';
```
Comment (paninaro) [inline src/components/app/App.tsx:18]: Will remove

**Review by seeden — APPROVED** [review]
LGTM

**Pull request merged**